cmd/collectors/rest/plugins/shelf/shelf.go

```diff
@@ -0,0 +1,303 @@
+					my.Logger.Info().Msgf("added instance label: (%s) [%s]", attribute, display)
+				}
+			} else {
+				metric, err := my.data[attribute].NewMetricFloat64(metricName)
+				if err != nil {
+					my.Logger.Error().Stack().Err(err).Msg("add metric")
+					return err
+				}
+				metric.SetName(display)
+				my.Logger.Info().Msgf("added metric: (%s) [%s]", attribute, display)
+			}
+		}
+
+		my.Logger.Info().Msgf("added data for [%s] with %d metrics", attribute, len(my.data[attribute].GetMetrics()))
+
+		my.data[attribute].SetExportOptions(exportOptions)
+	}
+
+	my.Logger.Info().Msgf("initialized with data [%d] objects", len(my.data))
+	return nil
+}
+
+func (my *Shelf) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
+
+	var (
+		records []interface{}
+		content []byte
+		err     error
+	)
+
+	// Set all global labels from rest.go if already not exist
+	for a := range my.instanceLabels {
+		my.data[a].SetGlobalLabels(data.GetGlobalLabels())
+	}
+
+	href := rest.BuildHref("", "*", nil, "", "", "", "", my.query)
+
+	err = rest.FetchData(my.client, href, &records)
+	if err != nil {
+		my.Logger.Error().Stack().Err(err).Str("href", href).Msg("Failed to fetch data")
+		return nil, err
+	}
+
+	all := rest.Pagination{
+		Records:    records,
+		NumRecords: len(records),
+	}
+
+	content, err = json.Marshal(all)
+	if err != nil {
+		my.Logger.Error().Err(err).Str("ApiPath", my.query).Msg("Unable to marshal rest pagination")
+	}
+
+	if !gjson.ValidBytes(content) {
+		return nil, fmt.Errorf("json is not valid for: %s", my.query)
+	}
+
+	results := gjson.GetManyBytes(content, "num_records", "records")
+	numRecords := results[0]
+	if numRecords.Int() == 0 {
+		return nil, errors.New(errors.ERR_NO_INSTANCE, "no "+my.query+" instances on cluster")
+	}
+
+	my.Logger.Info().Msgf("fetching %d shelf counters", numRecords.Int())
+
+	var output []*matrix.Matrix
+
+	// Purge and reset data
+	for _, data1 := range my.data {
+		data1.PurgeInstances()
+		data1.Reset()
+	}
+
+	results[1].ForEach(func(shelfKey, shelf gjson.Result) bool {
+
+		if !shelf.IsObject() {
+			my.Logger.Warn().Str("type", shelf.Type.String()).Msg("Shelf is not object, skipping")
+			return true
+		}
+
+		shelfName := shelf.Get("name").String()
+		shelfId := shelf.Get("uid").String()
+
+		for attribute, data1 := range my.data {
+			if statusMetric := data1.GetMetric("status"); statusMetric != nil {
+
+				if my.instanceKeys[attribute] == "" {
+					my.Logger.Warn().Msgf("no instance keys defined for object [%s], skipping", attribute)
+					continue
+				}
+
+				if childObj := shelf.Get(attribute); childObj.Exists() {
+					if childObj.IsArray() {
+						for _, obj := range childObj.Array() {
+
+							// This is special condition, because child records can't be filterable in parent REST call
+							// frus type can be [module, psu] and we would only need psu for our use-case.
+							if attribute == "frus" && obj.Get("type").Exists() && obj.Get("type").String() == "module" {
```

Comment:
we do support filters in framework. You can try below.

api/storage/shelves?return_records=true&fields=*&query_fields=frus.type&query=psu

Reply:
filter and query can support main object but not able to handle the child array object.